ansible/roles/ci/code_style_check/license/files/gather_missing_licences.py

```diff
@@ -0,0 +1,149 @@
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument(
+        '--path',
+        type=str,
+        required=True,
+        help='The path to the repo to check the licences.')
+
+    args = parser.parse_args()
+
+    with open('/tmp/git_source', 'r') as tmp_file:
+        source = tmp_file.read().strip()
+    return Data(args.path, source)
+
+
+def get_changes_in_pr(data):
+    """Takes in the data class and uses it to get the differences in the pr. It uses subprocess to
+       get all of the changes using github cli (gh). Then gets all the files changed by getting
+       a list of all strings containing '+++' or '---'."""
+    command = ["git", "branch"]
+    master_branch = subprocess.run(command, text=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    devel_branches = ""
+    for branch in master_branch.stdout.split("\n"):
+        if "devel" in branch:
+            devel_branches = branch.strip()
+
+    command = ["git", "diff", devel_branches, data.source]
+    with subprocess.Popen(command, text=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE) as process:
+        out, err = process.communicate()
+        if process.returncode != 0:
+            print(f"Failed to get changes:\nstdout: {out}\nstderr: {err}")
+            sys.exit(1)
+
+    for line in out.splitlines():
+        if "+++" in line:  # Changed files are marked with +++.
+            file_changed = line.split("+++")[1][3:]  # Gets the filename
+            file_path = os.path.join(data.path, file_changed)
+            _, extension = os.path.splitext(file_path)
+            if not extension:  # We sometimes have python files without extensions
+                with open(file_path, 'r') as code_file:
+                    firstline = code_file.readline().strip()
+                    if "#!/usr/bin/env python" in firstline:  # Can use to tell if py file
+                        payload = (file_path, "py")
+                        if payload not in data.changed_files:
+                            data.changed_files.append(payload)
+            elif extension[1:] in data.accepted_extensions:
+                payload = (file_path, extension[1:])
+                if payload not in data.changed_files:
+                    data.changed_files.append(payload)
+
+        if "---" in line:  # Changed files are marked with +++.
+            file_changed = line.split("---")[1][3:]  # Gets the filename
+            file_path = os.path.join(data.path, file_changed)
+            _, extension = os.path.splitext(file_path)
+            if not os.path.exists(file_path):  # For if files are removed.
+                continue
+            if not extension:  # We sometimes have python files without extensions
+                with open(file_path, 'r') as code_file:
+                    firstline = code_file.readline().strip()
+                    if "#!/usr/bin/env python" in firstline:  # Can use to tell if py file
```

Comment:
I would combine this with the branch name. As in if it's noetic-devel or a child of noetic-devel then it should be python3 otherwise python is enough

Reply:
Thats why i just look for env python, because it will find it no matter what, noetic-devel or earlier

Reply:
But a shebang of python3 in kinetic would satisfy your condition right? And you don't want that do you?

Reply:
Thats true, but this would be caught out in the other tests like code_style and install checks wouldn't it, it wouldn't make sense to add this to a test about licence. We just have this bit of code to check python files without extensions

Reply:
All right, nevermind then :)